### PR TITLE
fix | Implement de_ignored_any to consume unknown fields

### PR DESCRIPTION
**Pull request by fruno-bulax**
# Short Context

The MicroG implementation of FIDO2 includes the optional [transports](https://w3c.github.io/webauthn/#dictionary-credential-descriptor) which [isn't in ctap-types](https://github.com/solokeys/ctap-types/blob/7fa0f08b1fcb0ad9ba6ab08edd8e2b138f96f6e1/src/webauthn.rs#L84) and the "appid" extension.

After I day I tracked the issue down to the cbor parser, which fails on map fields not present in the struct it's trying to parse.

# Issue

When deserializing a map that has fields not present in the struct, the key is consumed, but the value is instead skipped entirely.

When deserialization continues, there is leftover unexpected data in the input, leading to errors (usually bad major)

# Fix

Implement `deserialize_ignored_any` for `Deserializer`, so the values are fully consumed.

This does kind of go against not supporting `deserialize_any`, but I don't think there's a way around it.

### CODE REVIEW

File: src/de.rs [hunk 1]
```diff
@@ -701,8 +701,17 @@ impl<'de, 'a> de::Deserializer<'de> for &'a mut Deserializer<'de> {
     where
         V: Visitor<'de>,
     {
-        // Ignore extra fields/options
-        visitor.visit_none()
+        match self.peek_major()? {
+            0 => self.deserialize_u32(visitor),
+            1 => self.deserialize_i32(visitor),
```
Comment (sosthene-nitrokey) [inline src/de.rs:706]: This means that a value that does not fit in a `i32` or a `u32` de-serialization will fail.
I believe this should not actually de-serialize the value but just discard it.

File: src/de.rs [hunk 2]
```diff
@@ -1028,6 +1037,32 @@ mod tests {
...
+
+        let mut buf = [0u8; 64];
+        let base = Struct { a: 22, b: -161 };
+        let sup = SuperStruct { a: 22, b: -161, c: base.clone() };
+
+        let ser = cbor_serialize(&sup, &mut buf).unwrap();
+        let de: Struct = cbor_deserialize(&ser).unwrap();
+
+        assert_eq!(de, base);
+    }
```
Comment (sosthene-nitrokey) [inline src/de.rs:1064]: Here if you replace the `u8` with `u64` and `22` with `0xFFFFFFFFF` this will lead to `DeserializeBadU32`.

**Review by sosthene-nitrokey — CHANGES_REQUESTED** [review]
Thank you for the pull request.

However I believe that `ignored_any` should completely ignore the value and end on a `visit_unit`. For example that is what `serde_json` does: https://docs.rs/serde_json/latest/src/serde_json/de.rs.html#1897-1903

**Comment by fruno-bulax** [comment]
> Thank you for the pull request.
> 
> However I believe that `ignored_any` should completely ignore the value and end on a `visit_unit`. For example that is what `serde_json` does: https://docs.rs/serde_json/latest/src/serde_json/de.rs.html#1897-1903

Seems you're right. This was pretty far outside my wheelhouse so I expected to screw something up :)

As you've already got a [better PR](https://github.com/nickray/cbor-smol/pull/6) out and are in a better position to maybe actually get it merged I'd be happy to just close this one.

**Comment by sosthene-nitrokey** [comment]
I will close this PR in favour of https://github.com/trussed-dev/cbor-smol/pull/6

Now that this has been moved to `trussed-dev` we (Nitrokey) will be able to merge and release the other PRs.

Thank you.

**Pull request closed without merging**